features/admin-file-attributes.feature

```diff
@@ -25,7 +38,18 @@
     And I should see a button input "Delete"
 
   @ok @issue-457 @javascript
-  Scenario: Sign in as admin to delete attribute
+  Scenario: Sign in as admin, delete an attribute of a published dataset and check history tab
+    Given I sign in as an admin
+    And I am on "/adminFile/update/id/13973"
+    And I should see "last_modified"
+    When I press "Delete"
+    And I go to "dataset/100056"
+    Then I should see "Termitomyces sp. J132 fungus genome assembly data."
+    And I follow "History"
+    And I should see "History" tab with text "Termitomyces_assembly_v1.0.fa.gz: file attribute deleted"
+
```

Comment:
Hi @rija 

I found this way `I should see History tab with text` to make sure that the attribute has been actually deleted for `Published` dataset.  
I will try to implement this `And I should see a dataset log entry containing "file attribute deleted"`, do you have any idea that I can access the `dataset_log` table in `context` file?

Reply:
Hi @kencho51,
I might have answered that to you on Monday already, but to record an answer for the question here, here's the clarification: The history tab displays entries from the dataset_log table, so it's enough to check there is an entry in the history tab using browser navigation (and the step sentence you've come up with for that is fine - although the "text" ending could be dropped or replaced with "entry"). There is no need to access the database table programmatically in the step implementation. Also it defeat the purpose of acceptance testing. 
Think of it as "Black Box Testing", whereas Unit testing is "White Box testing" (you can look those up).
It's ok and often necessary to interact with low-level elements like database or filesystem in "Given ..." steps, because they set up the scene for playing the feature that need to be verified with browser navigation.

Reply:
Hi @rija 

I must have forgotten what you said 😞 
Right now I am working on to pass the `curation_log` test, here is my scenario:
```
  Background:
    Given Gigadb web site is loaded with production-like data

    @wip @issue-#457 @javascript
    Scenario: Go to a non published dataset found in production-like database
      Given I sign in as an admin
      And I go to "/adminFile/update/id/xxx"
      And I should see some file attributes
      When I press "Delete"
      And I should not see file attributes
      And I go to "dataset/YYY"
      And I follow "History"
      Then I should not see some file attributes
```
But I think the above scenario is not a proper way to test.
Because,  I could not identify the correct `xxx` and `YYY`. And even I found the `xxx`, this step `And I go to "dataset/YYY"` will not pass as error `The dataset does not exist` will be found.

So, can I have your suggestion on how to pass the `curation_log` test?

Reply:
Hi @kencho51,

Among a dataset's attributes in the database table there are ``id``, and ``identifier``.

Controller actions that alter the dataset expect to have the internal id ``id`` as parameter (that's your xxx) as that's the most appropriate one to uniquely identify a row in a database table.

In ``/adminFile/update/id/xxx``, ``/id/`` is how yii framework knows that we are looking for a dataset whose ``id`` attribute is ``xxx``

``identifier`` is for the DOI and is the public identifier for the dataset, that's why it is used in the view as the dataset url is meant to be shared publicly (that's your ``yyy``).

So when you pick a dataset as example, select both attributes from the table.
```
gigadb=# select id, identifier, title from dataset;
id  | identifier |  title
...
 189 | 100133     | Reference data for phage M13 dsDNA generated with the Oxford Nanopore MinION.
 204 | 100143     | Software and supporting material for "SMAP: a streamlined methylation pipelinefor bisulphite sequencing".
 210 | 100155     | Supporting data for "Sparse whole-genome sequencing identifies two loci for major depressive disorder".
...
```

Regarding your second point: datasets that are not published cannot be accessible publicly, so going to the public url ``/dataset/yyy`` would result in that error.

You've got couple of approaches:

You could add Given steps to change the upload status temporarily away from published, but then you would need to find a way to revert the change after the test has run.

Or you could have the Then steps to check this page:
(http://gigadb.gigasciencejournal.com:9170/datasetLog/admin/)
or more precisely manually navigate to the last page, and use that in a Then step.
It's not the most robust (the last page may change if we add/remove example datasets from production-like), but it's the easiest.

A better approach  is to use the test database instead of production-like then checking`` /datasetLog/admin/``'s first page would be enough, but you will need some Given steps to add some file attributes first to a dataset as I think the test database doesn't have any.

An improved alternative to that latter pre-condition would be to update the test data for the test database to have some file attributes in it so you don't have to add extra Given steps (I've pasted bash code to do that easily in @pli888's PR 532, you'd just have to update the corresponding CSV file beforehand)

Reply:
Hi @rija 

I finally passed the acceptance test on `curation log` by checking the last page of `datasetLog/admin`.
This dataset `100245` with `upload_status == Pending`, I manually add attribute to its file with `id == 95354` then delete it.

```
  @ok @issue-457 @javascript @NonPublished
  Scenario: Go to a non published dataset found in production-like database, create then delete a keyword attribute
    Given I sign in as an admin
    And I am on "/adminFile/update/id/95354"
    And I press "New Attribute"
    And I select "keyword" from "FileAttributes_attribute_id"
    And I fill in "FileAttributes_value" with "test Bauhinia"
    And I press "Add"
    And I should see "test Bauhinia"
    And I press "Delete"
    And I should not see "test Bauhinia"
    #Go to the last page of dataset log
    When I go to "datasetLog/admin/DatasetLog_page/74"
    Then I should not see "100245"
    And I should not see "FCHCGJYBBXX-HKBAUpcgEAACRAAPEI-201_L2_2.fq.gz: file attribute added"
    And I should not see "FCHCGJYBBXX-HKBAUpcgEAACRAAPEI-201_L2_2.fq.gz: file attribute deleted"
```

Reply:
It might be better to wait a bit longer for [PR #548](https://github.com/gigascience/gigadb-website/pull/548) to be merged then add some file attributes into the relevant CSV file test data. This will then allow the above scenario to be more concise since you wouldn't need to have the steps to add a file attribute for the test to delete.

Reply:
OK.

Reply:
PR #548 has now been merged into `develop` branch. If you integrate `develop` with your feature branch then you will be able to add the required test data into the relevant CSV file(s).

Reply:
This branch has been rebased with the `develop` branch, and attribute value `test Bauhinia` is added to the `data/production_like/file_attributes.csv`. So the following steps could be removed.
```
And I press "New Attribute"
And I select "keyword" from "FileAttributes_attribute_id"
And I fill in "FileAttributes_value" with "test Bauhinia"
And I press "Add"
```